Add eager parsing with clear indication of input

pragmaticanon opened the pull request:
This fixes #1185, with the following changes:
1. Make \ a separating delimiter, so when you hit enter after a trailing \, the cli will show >>> (this will fix 1 and 3)
2. Still parse each statement and throw error if there's a syntax error (this will fix 2)
3. Run all queries when a ; is entered at the end of any statement

eatonphil committed:
Add Go two-phase-many sample project

matklad committed:
testing: introduce TmpTigerBeetle

This is utility for integration tests, which gives you a temporary
TigerBeetle cluster.

matklad committed:
shell: echo commads which produce output anyway

matklad committed:
shell: autotrim newline

This matches bash `$()` behavior I believe.

matklad committed:
shell: workaround ziglang/zig#16779

matklad committed:
testing: echo TmpTigerBeetle stderr

matklad committed:
testing: workaround ziglang/zig#16820

matklad committed:
zig: protect from BoundedArray.len overflow

`@TypeOf(BoundedArray.len)` is the smallest integer that can hold the
array length. As such, even something like `array.len + 1` can overflow,
if (comptime) array len happens to be exactly power of two less one.

Protect against this by adding our own wrapper with a safer API.

The overall guideline here is:

- we wrap std types to fix API errors, or otherwise tweak them for our
  particular usage
- we do this lazily, only when there's an actual problem
- we do this minimally (wrap API rather than vendor the whole copy)
- we _might_ unfork once Zig stdlib is stable (and switches to
  snake_case :-))

matklad committed:
build: remove wrong cyclic dependency

unit-tests are a part of VSR package, they should directly poke internal
APIs via file imports, rather than exercising things strictly via public
API.

sentientwaffle committed:
VSR: Update outdated comments

sentientwaffle committed:
Merge pull request #1107 from tigerbeetle/dj-manifest-level-snapshot-order

LSM/ManifestLevel: Order tables by (key, snapshot_min) instead of (key)

sentientwaffle committed:
Merge pull request #1134 from tigerbeetle/dj-vsr-comment-fixes

VSR: Comment fixes

eatonphil committed:
Client built into the binary

cb22 committed:
go: Zig 0.10.1 -> 0.11.0

cb22 committed:
Merge pull request #1136 from tigerbeetle/cb22/lets-go-0.11.0

go: Zig 0.10.1 -> 0.11.0

matklad committed:
vsr: extra test for checksum

vsr.checksum is pretty important, as it cements our on-disk
representation and over-the-wire protocol (so, we won't be able to
easily change it).

This PR adds a "change detector" test to ensure we don't accidentally
change the result. The idea of the test is basically to hash a bunch of
random stuff, and then assert that the final hash has a specific value.

While we are at it, extend our fuzz test to do more sanity checking:

- allow arbitrary bitflips (rather than just +%= 1)
- also test avalanching (that every bit of input affects every bit of
  the output)

This actually uncovered  bug in the implementation: we were passing a
message as secret text to AEAD, but we should have passed it as
associated data, as the zig stdlib does.

Additionally, while this PR was in flight, we upgraded to Zig 0.11,
so switch to that implementation as well.

sentientwaffle committed:
VSR/Grid: Rename: Callback.from_local{,_or_global}{,_storage}

sentientwaffle committed:
VSR/Grid: Rename Writing.none to Writing.not_writing

To avoid "writing for an unknown reason" confusion.

sentientwaffle committed:
VSR/Grid: Document Grid.Read.coherent

sentientwaffle committed:
VSR/Grid: Fix: Don't remove cache entry unless we wrote to the cache

sentientwaffle committed:
VSR/Grid: Simplify read_block_callback control flow

matklad committed:
Merge pull request #1080 from tigerbeetle/matklad/checksum-checks

vsr: extra test for checksum

sentientwaffle committed:
VSR: Rename on_request_blocks_read_{repair,block}

matklad committed:
testing: upgrade tmp_tigerbeetle to 0.11

matklad committed:
testing: try another way to get at the root of the repo

* git rev-parse --show-top-level does not work on alpine
* cd .. does not work on windows
* @src().parent.parent does not work on Zig 0.11
* assuming that we run from the root does not work for clients

sentientwaffle committed:
VSR/Grid: Add ReadBlockResult caveat

eatonphil committed:
Fix tigerbeetle version: correct zero check in src/shell.zig, and git is a dependency during build steps

sentientwaffle committed:
VSR/Grid: Add missed checkpointing assertion

sentientwaffle committed:
VSR: Extract Op math helpers

sentientwaffle committed:
VSR/Replica: s/op_checkpoint_trigger/op_checkpoint_next_trigger

This clarifies some terminology about checkpoints:
- No need to refer to a checkpoint's "previous" or "next" (past/future) trigger.
- Instead, every checkpoint has exactly one trigger (except checkpoint=0, which has no trigger).
- This corresponds to the "past" trigger. To refer to the "next" trigger, compute the trigger of
the next checkpoint.

sentientwaffle committed:
VSR: Free up "Op" identifier in vsr.zig

sentientwaffle committed:
VSR/Op: Shorten function names

matklad committed:
message_pool: fix wrong alignment in free

We must pass the same alignment in free as the one used in AlignedAlloc.
Just removing the cast should fix the  issue. I considered spelling the
alignment here explicitly, but I think it's better not too -- we have a
bunch of similar cases elsewhere, and they all work fine. The problem we
have here is specifically due to us casting the alignment away with @as.

matklad committed:
Merge pull request #1140 from tigerbeetle/matklad/alligned-free

message_pool: fix wrong alignment in free

sentientwaffle committed:
VSR/Op: Code review

batiati committed:
Zig 0.11.0 changes

- Replace the pattern @ptrFromInt(@intFromPtr(...)) in many places, using `@constCast` or `@ptrCast`
- Remove unecessary `@as`, prefering explicity types declarations instead.

batiati committed:
Prefetch eytzinger

batiati committed:
Add isPowerOfTwo assert and update comments

batiati committed:
Merge pull request #1141 from tigerbeetle/batiati-constcast

Zig 0.11.0 misc

sentientwaffle committed:
VSR/Grid: Add explanation

sentientwaffle committed:
SuperBlockManifest: s/tree_hash/tree_id

sentientwaffle committed:
Merge pull request #1120 from tigerbeetle/dj-grid-read-repair

VSR/Grid: read_block(.from_local,{_or_global})

sentientwaffle committed:
LSM: Change tree ids from u128 to u16

Leave the superblock manifest formatting as-is for the moment, since tree ids will shorty be removed from it anyway.

eatonphil committed:
Merge pull request #1138 from tigerbeetle/pe/git-tag

Fix tigerbeetle version, correct zero check in src/shell.zig

sentientwaffle committed:
SuperBlockManifest: Line lengths (tidy)

sentientwaffle committed:
SuperBlockManifest: Remove unnecessary memset

Both `init()` and `reset()` already zero the SoAs – no need to re-zero them.
(`SuperBlock.Manifest.verify()` asserts this).

eatonphil committed:
Explicit check for prefix characters in release to use in Linux version check

batiati committed:
Merge pull request #1135 from tigerbeetle/batiati-eytzinger

Add prefetch to Eytzinger

batiati committed:
Simplify the block layout.

- Deletes `eytzinger.zig` and removes the layout block, along with key_layout* constants and derived configurations.

- Only computes bloom filters where `Table.usage == .general`, although still keeping the filter block on the layout.
Asserts that filter blocks are only used for `Table.usage == .general`.

- Repurposes the `eytzinger_benchmark` as a `binary_search_benchmark`.

- Refactors the `// TODO u16 + padding.` at `schema.zig` for `TableData.Parent`.

sentientwaffle committed:
Merge pull request #1139 from tigerbeetle/dj-vsr-op-helpers

VSR: Extract Op helpers; rework "checkpoint trigger" terminology

batiati committed:
Fix failing fuzz tests

sentientwaffle committed:
VSR/Grid: s/read_faulty_queue/read_global_queue

This way the queue matches `from_local_or_global_storage`.
Also, with the repair queue, "faulty" is not a precise description of this queue.

sentientwaffle committed:
VSR: commit_unsynced_{min,max}

When state sync completes, the replica still needs to repair the LSM table blocks and VSR client replies that it skipped over.

How it works:

As part of the new superblock installed by state sync, set `vsr_state.commit_unsynced_{min,max}`.
(If `commit_unsynced_max ≠ 0` _before_ this sync, it means we were still repairing from a previous sync. In that case, we grow to range to encompass both sync's commits.)

On every checkpoint, either:
- If all of the tables/replies are synced, set `commit_unsynced_{min,max}` to zero. (See `sync_content_done()`).
- If all of the tables/replies _aren't_ synced, reuse the same values for `commit_unsynced_{min,max}`.

When we open the state machine, if `commit_unsynced_max ≠ 0`, start syncing replies and tables.

`commit_unsynced_max`:

    /// When zero, all of the grid blocks and replies are synced.
    /// (When zero, `commit_unsynced_min` is also zero.)
    ///
    /// When nonzero, we must repair grid-blocks/client-replies that would have been written
    /// during the commits between `commit_unsynced_min` and `commit_unsynced_max` (inclusive).
    /// (Those grid-blocks and client-replies were not written normally because we "skipped"
    /// past them via state sync.)

matklad committed:
Merge pull request #1114 from tigerbeetle/matklad/tmp-tigerbeetle

testing: introduce TmpTigerBeetle

sentientwaffle committed:
VSR: Sync client replies

sentientwaffle committed:
Test/StorageChecker: Check client replies

(Grid checking is still disabled, until table sync is done.)

sentientwaffle committed:
Test/Storage: area_memory()

sentientwaffle committed:
Test/StorageChecker: Remove compaction hook

Remove this hook for the moment, since it won't work until the unified manifest is ready.

eatonphil committed:
Add node two-phase many sample project

eatonphil committed:
Merge pull request #1051 from tigerbeetle/pe/two-phase-many

Add Go sample project for many in-flight two-phase transfers

sentientwaffle committed:
VSR: SuperBlock.replica_index

sentientwaffle committed:
VSR/SuperBlock: Log replica index

batiati committed:
Merge pull request #1145 from tigerbeetle/batiati-remove_eytzinger

Simplify the block layout.

batiati committed:
Implements @prefetch to binary search accordingly to https://arxiv.org/abs/1509.05053.

No divergences from the original paper, except that in our case a Value can be larger than a single cache line, and we need to prefetch multiple addresses for a single value.

Prefetch is opt-out (e.g. .{ .prefetch = false }).
Benchmark ./zig/zig build benchmark_binary_search compares both versions.

matklad committed:
zig: fully drop previously vendored HashMap

sentientwaffle committed:
VSR/SuperBlock: Set replica_index in format() too

sentientwaffle committed:
VSR/Grid: Document read_global_queue

sentientwaffle committed:
Merge pull request #1146 from tigerbeetle/dj-grid-global-queue

VSR/Grid: s/read_faulty_queue/read_global_queue

sentientwaffle committed:
Merge pull request #1148 from tigerbeetle/dj-superblock-replica-log

VSR/SuperBlock: SuperBlock.replica_index

batiati committed:
Merge pull request #1144 from tigerbeetle/batiati-binarysearch

Prefetch binary search

matklad committed:
Merge pull request #1149 from tigerbeetle/matklad/stdx-hash-map

zig: fully drop previously vendored HashMap

eatonphil committed:
Merge pull request #1142 from tigerbeetle/pe/fix-for-version-check

Explicit check for prefix characters in release to use in Linux version check

matklad committed:
Merge pull request #1122 from tigerbeetle/matklad/bounded-array-len

zig: protect from BoundedArray.len overflow

batiati committed:
Refactor tree_id to u16.

batiati committed:
Merge pull request #1151 from tigerbeetle/batiati-refactor

Refactor tree_id to u16.

matklad committed:
wip: docs: what we store on disk

eatonphil committed:
Add dotnet two-phase many sample project

matklad committed:
Merge pull request #1130 from tigerbeetle/matklad/no-vsr-for-unit

build: remove wrong cyclic dependency

eatonphil committed:
Add java two-phase many sample project

kprotty committed:
node: build_lib.js which generates node.lib

Moves/replaces the `windows_generate_node_lib.js` script
for creating the `node.lib` file needed for `zig build node_client`
to work on windows. This also modifies `build.zig` to actually create
`tb_nodeclient.dll` for windows.

sentientwaffle committed:
VSR/Grid: Remove invalid assert

This assert doesn't work since latent read errors mean that a read may fault, then another read to the same block can succeed (populating the cache).

sentientwaffle committed:
VSR: s/commit_unsynced_/sync_op_

sentientwaffle committed:
Docs: Distinguish "sync superblock" and "sync content" protocols

cb22 committed:
ci: Fix Docker push step

cb22 committed:
Merge pull request #1155 from tigerbeetle/fix-docker-push

ci: Fix Docker push step

eatonphil committed:
Merge pull request #1147 from tigerbeetle/pe/node-two-phase-many

Add Node sample project for many in-flight two-phase transfers

eatonphil committed:
Add roll forward setting to .NET sample projects

eatonphil committed:
Merge pull request #1153 from tigerbeetle/pe/dotnet-two-phase-many

Add Dotnet sample project for many in-flight two-phase transfers

kprotty committed:
node: create client.node from all dist/bin files

The prepack script assumed all binaries were `.so` files and failed CI
on the new `.dll` being built. This scans and renames both type of files,
including `.dylib`s.

kprotty committed:
Merge remote-tracking branch 'origin/main' into king-node-windows

sm2n committed:
lsm: cleanup loop

sm2n committed:
lsm: Remove unnecessary intCast

sm2n committed:
lsm: Assert data block schema vs vsr.Header.size and padding_offset

sentientwaffle committed:
VSR: Ease checkpoint op restrictions

LSM doesn't care about the checkpoint interval, VSR doesn't need to either.
In the future we may want to allow user/control-protocol driven checkpoints.
(For example, we might want to immediately before a restart of upgrade to minimize recovery time).

sentientwaffle committed:
Lint: fmt

sentientwaffle committed:
Merge pull request #1159 from tigerbeetle/dj-vsr-ease-checkpoint-requirements

VSR: Ease checkpoint op restrictions

kprotty committed:
Merge pull request #1154 from tigerbeetle/king-node-windows

Node: support building `tb_nodeclient.dll`

eatonphil committed:
Add note about oomkiller to docker-compose docs

sentientwaffle committed:
VSR: Don't clear faulty replies that are still faulty

sentientwaffle committed:
VSR/SuperBlock: Add commit_min_canonical

This is a step towards avoiding unnecessary re-sync of data, while still tolerating checkpoint divergence.

sentientwaffle committed:
Docs: s/SuperBlock/Superblock when referring to concept rather than implementation

Context: https://github.com/tigerbeetle/tigerbeetle/pull/1160#discussion_r1304593387

sentientwaffle committed:
VSR: More specific sync_content_done

matklad committed:
tie things together

sm2n committed:
lsm: cleanup loop

sm2n committed:
lsm: under constants.verify, verify that read data blocks are sorted

sentientwaffle committed:
VSR/SuperBlock: Fix grid_blocks_max, data_file_size_min

`grid_blocks_max` and `data_file_size_min` were not counting the client replies.

(This was missed because we don't have any tests that fill the whole grid.)
(We don't test a full grid yet since it isn't currently handled properly.)

Also, in `replica_test.zig`, use `storage_size_limit == storage_size_max`.
This is needed because a test (`Cluster: recovery: grid corruption (disjoint)`) iterates over every block in the grid and marks faults.

sentientwaffle committed:
RingBuffer: Runtime capacity for buffer_type=slice

Replace buffer_type=pointer with buffer_type=slice.
Use a runtime capacity instead of comptime.

This makes some option plumbing easier farther up the stack.
And it also minimizes specialization (though that probably doesn't matter here since most of `RingBuffer`'s functions are inlined. And also we don't use buffer_type=slice/pointer anywhere else.)

sentientwaffle committed:
RingBuffer: Add RingBuffer.get()

sentientwaffle committed:
NodePool: Use BitSet.setRangeValue() for reset()

sentientwaffle committed:
LSM: refAllDecls

eatonphil committed:
Merge pull request #1156 from tigerbeetle/pe/java-two-phase-many

Add Java sample project for many in-flight two-phase transfers

sentientwaffle committed:
VSR: s/journal_size_max/journal_size

The journal size is constant.

sentientwaffle committed:
LSM: Unified manifest

## Summary

- Store tree id in TableInfo in the manifest log blocks on disk.
- Store TableInfo of every tree in each manifest log block.
- Move the `ManifestLog` from one-per-`Tree` to one-per-`Forest`.

This change still maintains the original freeset reserve/forfeit pattern (adapted for the forest-level manifestlog). (This is not optimal but will be amended in a future commit.)

#### Results

- Worst-case manifest log writes for a compaction half-bar: 63 blocks to 6 blocks.
- Worst-case manifest log writes for a checkpoint: 63 blocks to 6 blocks.
- Average case manifest log writes for a checkpoint: 21 blocks to 2-3 blocks. (This is a ballpark guess).
- Memory owned by manifest log: 63 blocks to 6 blocks.
- `state_machine.checkpoint()` latency drops by ~70% in the benchmark. (Note that this is an _extremely_ small sample size, though.)

<details>
<summary>Checkpoint latency samples</summary>
(Measuring `state_machine.checkpoint()` from invocation to callback.)

main:
0.041102385s
0.044551919s
0.033833146s
0.033007621s

unified manifest:
0.010172937s
0.012081376s
0.009965735s
</details>

## Context

Each tree's `TableInfo` is currently (abbreviated):

    fn TableInfoType(Key) type
        return extern struct {
            checksum: u128,
            address: u64,
            flags: u64, // (Currently unused.)
            snapshot_min: u64,
            snapshot_max: u64,
            key_min: Key,
            key_max: Key,
        };
    }

Some relevant constants:

    block_size                             = 64KiB
    lsm_batch_multiple                     = 64 ops
    journal_slot_count                     = 1024 ops
    storage_size_max                       = 16TiB
    tree_count                             = 21 trees
    lsm_levels                             = 7
    bars_per_checkpoint                    = 16 = 1024 / 64

    // Worst-case: fully-loaded LSM, with every possible compaction running every bar:
    compactions_pre_bar_per_tree           = 112  = lsm_levels * bars_per_checkpoint
    compactions_per_bar                    = 2352 = lsm_levels * bars_per_checkpoint * tree_count

    // This is a pretty arbitary constant – we can tune it up or down (preferrable down) as necessary.
    // But it is part of the disk layout, so we need to pick a value and stick with it.
    // (See superblock_trailer_manifest_size_max.)
    maximum superblock manifest size       = 36_184_064B = ~34.5MiB

And currently:

    SuperBlock.Manifest.BlockReferenceSize = 40B = @sizeOf(tree_id) + @sizeOf(checksum) + @sizeOf(address)

    // Pretending that every manifest log block is full (which they aren't in practice – see below):
    maximum manifest log blocks            = ⌊36_184_064B / 40B⌋ = 904_601 blocks
    maximum manifest log bytes             = 904_601 * block_size = ~55.2GiB
    // Pretending that all Keys are 16B (as they are for the majority of trees).
    maximum manifest updates               = 723_680_800 updates = 800 updates/block * 904_601 blocks

    Trees:
    | @sizeOf(Tree.Key) | @sizeOf(TableInfo) | manifest updates/block | Tree Count |
    |                8B |                64B |                    992 |          4 |
    |               16B |                80B |                    800 |         12 |
    |               32B |               112B |                    576 |          5 |

However, the _updates-per-block_ numbers (and thus _maximum manifest updates_) above are a misleading ballpark:

- Different trees have different manifest log blocks.
- Different levels of the same tree have the same manifest log blocks.
- Lower bound: the LSM writes one manifest log block per tree (that is written to during the checkpoint) per checkpoint. (21 trees → 21 blocks written/checkpoint).
- Not counting manifest log compaction, each of those blocks has _at most_ 112 (`compactions_pre_bar_per_tree`) blocks – a write/space amplification of ~7 (`TableInfo/block ÷ compactions_pre_bar_per_tree`).
- Usually we don't compact every tree of every level. So the blocks are even emptier! (More space/write amplification).
- The amplification isn't a huge deal, but those 21 writes are a lower bound, and they occur during the checkpoint, contributing to the checkpoint latency spike.
- Additionally, the incomplete manifest log blocks increase the amount of manifest log compaction work outside of checkpoints.

## Unified Manifest Log

Instead of dividing the manifest log between trees, we could put all updates to all trees in the same manifest log blocks.

For this to work, `TableInfo` would need:

- To include a tree identifier (recently migrated from `u128` to `u16`).
- To accommodate any table, the `Key`s must be padded to the largest key size of any tree (`32B`).
- Then `@sizeOf(TableInfo) == 112B` _for every tree_.
- Lower bound: the LSM writes one manifest log block total per checkpoint.

On the superblock manifest side:
- No need to store the `tree_id` in the superblock manifest anymore, so `SuperBlock.Manifest.BlockReferenceSize == 24B` (down from `40B`).

Then:

    // Using the same value as a base, then working backwards:
    maximum manifest updates               = 723_680_800 updates
    maximum manifest log blocks            = 722_680_800 updates ÷ 576 updates/block = 1_256_390 blocks

    // 24B = @sizeOf(checksum) + @sizeOf(address)
    SuperBlock.Manifest.BlockReferenceSize = 24B
    maximum superblock manifest size       = 1_256_390 blocks * 24 bytes/block = 30_153_360B = 28.8MiB

This means that despite paying for the largest key size for trees with small keys, the overall superblock manifest size shrinks as we drop the tree id.
(Note that this is because we currently encode tree id as a `u128` even though it fits in a `u16`.)

#### Pros

- Smaller superblock manifest upper bound.
- Smaller superblock manifest in practice (due to less space amplification).
  - This means less data for checkpoints to write.
  - This means a little less data for state sync to move.
- Lower manifest log space/read/write amplification.
  - This reduced the amount of work necessary when starting the replica (i.e. opening the LSM).
- Fewer manifest log writes, especially at checkpoint.
- Faster recovery (fewer, denser manifest blocks to read).
- Lower `ManifestLog` RAM usage (approximately `21*3*64KiB ≈ 4MiB` → `6*64KiB = 384KiB`).
- De-specialize `ManifestLog`.

#### Cons

- We can't add a tree with a key larger than `32B`.
  - (Well, we can, but we would need to re-version the manifest block format).
  - (For secondary indexes, that means the maximum is `8B timestamp + 24B indexed key`).
  - (If we anticipate one a larger key, we could pad in advance, but the above calculations all assume 32B is the limit.)
- Unified manifest log compaction can't separate hot and cold manifests – all of the entries in a manifest log block are rewritten when it is compacted.
  - (So if one tree is modified infrequently, while the other is modified often, rewriting the cold tree's manifest entries is now write amplification).

<details>
<summary>Unified Manifest Level?</summary>
## Unified Manifest Level?

All of the above is entirely about the manifest _on disk_.

For completeness: **What about the in-memory manifest?**

Two options:
1. Specialized. (This is what we already have implemented.)
2. Unified.

I lean towards _specialized manifest levels_.
But either way, the decision doesn't impact the storage format, so it is flexible in the future.

### Specialized

Right now we have per-tree `TableInfo`, with a per-tree specialized `ManifestLevel`.

#### Pros
- This is more RAM-efficient – no overhead due to padding keys up to 32B.
- (Also, this is already implemented.)

#### Cons
- We need separate `TableInfo` type(s) from the `TableInfo` that is serialized into the manifest log blocks. (`schema.TableInfo` vs `ManifestLevel(Tree).TableInfo`)
- Per-tree specialization of `ManifestLevel` (i.e. more code generation).

### Unified

Change `ManifestLevel` (and `SegmentedArray`) to take a `*const fn` for `compare_keys` (instead of a `comptime fn callconv(.Inline)`).

#### Pros
- No pre-tree specialization of `ManifestLevel` (i.e. less code generation).

#### Cons
- More RAM overhead – tables with 8B or 16B keys must still pad their keys to 32B.
- `compare_keys()` can't be inlined.
- My guess is that despecializing `SegmentedArray` will have a noticeable (CPU) performance impact when searching levels with many tables.

</details>

sentientwaffle committed:
SuperBlock/Manifest: Assert removed from compaction-set

sentientwaffle committed:
SuperBlock/Manifest: s/BlockReferenceSize/BlockReference.size

Don't CamelCase integer constants.

sentientwaffle committed:
LSM/Forest: Simplify "Join"

Only compaction needs to join.
Also reduce type state dimensionality by grouping multiple correlated optional fields into `?struct {...}`.

sentientwaffle committed:
LSM/Groove: Simplify "Join"

Only compaction needs to join.

sentientwaffle committed:
LSM: Fix Forest/Groove fieldParentPtr alignment

matklad committed:
vsr: use stable checksum for ClusterConfig

This checksum could leak to disk or network, so it must be stable across
releases of TigerBeetle.

Closes: #1009

sentientwaffle committed:
Merge pull request #1160 from tigerbeetle/dj-vsr-sync-op-superblock

VSR: State Sync (Part 3/?): sync_op_{min,max}; sync replies

sentientwaffle committed:
Merge pull request #1162 from tigerbeetle/dj-superblock-size-with-replies

VSR/SuperBlock: Fix grid_blocks_max, data_file_size_min

matklad committed:
mention how the tables are limited in size

sentientwaffle committed:
LSM/ManifestLog: Refactor reserve/forfeit

- Move the ManifestLog's `reserve()` to be internal.
- Name ManifestLog's `forfeit` `compact_end()` to match the trees.
- In Forest, perform the ManifestLog's `reserve()` after the Tree's.
  - (This doesn't matter either way, I just liked it better this way.)

I went down a rabbit hole trying to split ManifestLog's reserve/forfeit into two separate staged (ML-compaction and tree-updates), in such a way that the latter would reserve exactly as many addresses as it needed, to avoid any space amplification. That worked, but it increases the ML's API surface area and makes its usage more awkward -- for little benefit -- so I rolled that back.

sentientwaffle committed:
LSM/ManifestLog: Add misc state assertions

sentientwaffle committed:
VSR/SuperBlockFreeSet: Remove count_free_reserved()

Function isn't used anymore.

sentientwaffle committed:
LSM/ManifestLog: Rename Options.tree_count to Options.forest_tree_count

sentientwaffle committed:
LSM/ManifestLog: Concurrent writes instead of sequential

sentientwaffle committed:
LSM/ManifestLog: Tidy

rbino committed:
bootstrap.ps1: make curl follow redirects

batiati committed:
Merge pull request #1164 from rbino/follow-redirects-in-windows-bootstrap

bootstrap.ps1: make curl follow redirects

matklad committed:
Merge pull request #1163 from tigerbeetle/matklad/checksum-config

vsr: use stable checksum for ClusterConfig

matklad committed:
lsm: remove useless cast

Just making sure compiler catches us out if the types change here

matklad committed:
lsm: assert that data block padding is zero

matklad committed:
lsm: drop meaningless assert

key_max just _is_ `values[values.len - 1]`, so this isnt' asserting
much. Will add against the sentinel key shortly.

matklad committed:
lsm: finalize data block before any asserts and mutations

matklad committed:
lsm: group block assertions

matklad committed:
lsm: cosmetics

matklad committed:
lsm: assert they keys are smaller then the sentinel

sentientwaffle committed:
LSM/ManifestLog: Assert relationship between table-count and manifest-log-block-count

The amplification factor is arbitrary (other than that it must be an integer greater than 1), unfortunately.

eatonphil committed:
Merge pull request #1161 from tigerbeetle/pe/docker-compose-docs-error

Add note about oomkiller to docker-compose docs

cb22 committed:
Fix CI benchmark client args

matklad committed:
build: use builtin fmt step

matklad committed:
Merge pull request #1165 from tigerbeetle/matklad/lsm-padding-assert

lsm: assert that data's block padding is zero

sm2n committed:
stdx: Remove zeroed performance todo because it's already fast

I tried a couple explicit simd versions but it turns out that in
ReleaseSafe the naive thing generates basically the same code as far
as the best version as far as I can tell.

Also, u128s are atrocious apparently. For some reason the
autovectorizer doesn't kick in at all.

also a s/u64/u128/g version of this

pub fn zeroed(bytes: []const u8) bool {
    var byte_bits: u64 = 0;
    const bytes_split = bytes.len - @mod(bytes.len, 8);
    for (std.mem.bytesAsSlice(u64, bytes[0..bytes_split])) |group| {
        byte_bits |= group;
    }
    for (bytes[bytes_split..bytes.len]) |byte| {
        byte_bits |= @as(u64, byte);
    }

    return byte_bits == 0;
}

pub fn zeroed(bytes: []const u8) bool {
    var mask: @Vector(16, u8) = @splat(0);
    const bytes_split = bytes.len - @mod(bytes.len, 16);
    var i: u32 = 0;
    while (i < bytes_split) : (i += 16) {
        const vec: @Vector(16, u8) = bytes[i..][0..16].*;
        mask |= vec;
    }

    var byte_bits = @reduce(.Or, mask);

    for (bytes[bytes_split..bytes.len]) |byte| {
        byte_bits |= byte;
    }

    return byte_bits == 0;
}

Here's a godbolt:
https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AAvPMFJL6yAngGVG6AMKpaAVxYM9DgDJ4GmADl3ACNMYhAAJgA2UgAHVAVCWwZnNw89eMSbAV9/IJZQ8OiLTCtshiECJmICVPdPLhKy5MrqglzAkLDImIUqmrr0xr62jvzCnoBKC1RXYmR2DgB6JYBqABUAT1jMVc3Z4lW0LFWEMMxSVZJV2lQmdFXDVcxVVlj6ADoAUg0AQTQGH1Vn0Hl8AMwAEVWAAE8Cx4jUIF8IhEQciIpNwQAhH6/XG42KuYKrBioPy0Py7KgMVbGMKoTDoCDBTYETAKECrL4AVhx3IhAKBrgAHJNVsFUC4uQB2HF/VYK1YAN2q4tZmAA%2BsFCBzViKuZDVhpsbjFat%2BIdmeqFGKvtLHCy2XbHDK5b8zWbHZrtQQFDLHOCoV6TfLFXaIfjQwriJgCHMaV6tTqDRHDcawW7w5HfoTiaTyZTzTS6cQGeguFa2bqeXyBQIhaLxZLaK7TYqVYdEz7da4opIU0aQ%2B7FYKCGqqxqFO9CAOvQoPvQaQBaGEsDCV9kLxiXUVDs0W1YQEEfNgsD5z34KIQUhYQXvScfsmsaD7n62T6cEHkQya2%2B3AUtXFiZ1WyjD0u2TZ1A1WADZmAjM2wVLMwIPDcFBrOcPwpAhXznLcGG/P8HXVEC7TdD1FQg31/Wg6EmAUO8%2B0uL1MQQsDkLxMCYzjYgE3VJNqMDaD00zaUIz%2BAkiRJMkGApfwi1pelGQiNDORrb9Rz1RsJSlMjEOVVUqJ7LgImFAcRP0zTMKnbDZ2tfDVhXaE1yZPDF0uLgolY8jFVQ49TzfKtL2vPBb1cEzhWY61n1w98bMIQiiNgoDSNlfTPX47saMNZL4J8pCxP01C5wwuLP1iqt8MI/1g3tUDhwoozsqhOiGPC0yorZbz9I4/TuPjR8BL9IS0yHXq/lzaSC3k6lFNLRkwVUrleQ0%2BsxxFMUdJbPSwI7VYWHogBrTloQANUwawSAgTzLg2gdoRspgCAgDRurAqyyts6C3MYRzV3XH6GA8ry93bVUKD1MEInM0GFQAdwQOhdggPADRdazPzFTkUa5CIcUNTy/3yj1NKVC6TvOy7iGumItNtQ0St5PBX2/GKPk878PgAKlhj0DoUQ7muVC7efGhqDM7TLIMNaEY3QVxbw%2BAB5Mh9qOt7xaK640NKid4pwwLN0XarnVql0dvFjK2SGoXgzYzXCokrjYwGpqRqhCyJMdzifcm2aDr8CAxTAMAlTJUE0t21VgFiJgB2PM4mFiD4AHFGDCMRlDmTJMF%2BWhblEIhqe%2BNKxMxUvU2JrAqDCVYNQHGOmA%2BLA/EIIPec0sQC6e65oMbj4u9QQuro1yy1seYhgGGw0CGITZgQIdAPliUsFgUedqinvOC4gQfh%2BIDWzWr2vjxXof2Q3yeFAAMRjTBd/zoee9VzebTGsDjywYJXGAAf17CZ6JYyxHkXieTAZ4FA3lzgoLE1pd5X1/IfRUn9MDf1/vRJQiIgGMgrP5cBHxIGhWgbAqs8Cp6IN5igtBf9MGAKUugFSeCIFQMvCQ9kZCbS/koaAr%2BP8aEAIgNg9Ai0mEEJYTAuBr8KH22zGaKhfCV5%2BGeuiX4k93CMF9GpWU6FCrckcARFElwS5YlfuGUe3tswcGmLQTg3JeCeA4FoUgqBOAAC0zDAgOAsXGYIeCkAIJoKx0xjqSEkB8MEkgNBRA0FwaU3JhRRAAJwaGFJFGxHBJD2MCc4zgvAOQaH8YE6YcBYBIDQPCJGZAKAQHKbESpIBjCz1cAwQ6pAsBKiIadPAmA4ZKx2A4vxNBaBsmIByZk2TtTMDnpwPxkzqibCVsEbQF0AncF4OUtggglayU2Nk3hwBHBdw5Gstp4DDDAHEI43g%2BAYzWDwGTY5TiXgXVcGyGZvAlGlGyRSYIxB5nOCwNk2ecJ3nTCoAYKeXSel9MYO8mQggRBiHYFIeF8glBqGyboRoBgjANLMPoPAwQOSQGmKgWI5RjlLiVgAJVKJgeimAhBMBro5akqAlx9GIHgWIvBUBk2IFyrAxKg5NBWckewDAnAuHqBIbkpAfCUnGN0WJUQwRxASEkAQgwGhysyJqhgYwujhBVWqywYqBCtAGNKoYoq7kWv6O0RVRqJDSlVRYB12q9AjBqIagoyrXVgmmAoLxyLrG2KyVcnJHBaQeKaS0w8uBCC9wiL4yYvBVlaF/KQY6EREkfGFIkrgiTuQpo0NKCInkmLpMyaQBxTiXEcDySAApGarGkBKYgEAswCCEgIOQSgtTKkBFYIsUwwAjTsxfE8/ARc8DoD0PwBFohxAosXWilQ6hI1YtIHDP5sRQX6HDbW7JDalavN7VcKgMbx1xsFhAZwFT6CHGRKm9NRTgmRG5B8V10pQnckSREjQiSIiSGlIejJEb625IsM2wpVys3pIiJB3l0HW1Zv5YkOwkggA%3D%3D%3D

sentientwaffle committed:
Fuzz: Fix commit_min_canonical in superblock fuzzer

sentientwaffle committed:
LSM/Fuzz: Update fuzzers for unified manifest

matklad committed:
Merge pull request #1167 from tigerbeetle/matklad/fmt

build: use builtin fmt step

sentientwaffle committed:
Merge pull request #1168 from tigerbeetle/dj-fuzz-superblock-canonical

Fuzz: Fix commit_min_canonical in superblock fuzzer

batiati committed:
Node client support on Windows

Adds win32 to platformMap fixing the error Unsupported platform: win32.

Fixes the .lib file that was generated pointing to node.dll instead of
node.exe, causing the dlopen to fail trying to load node.dll that does
not exist.

Converts the sh scripts to js for cross-platform compatibility.

Adds windows-latest to the CI tests.

batiati committed:
Merge pull request #1170 from tigerbeetle/batiati-node-windows

Node client support on Windows

sentientwaffle committed:
LSM/ManifestLog: Don't block on manifest compaction until last beat

Start manifest log compaction on the first beat of each half bar.
Don't block on manifest compaction until the last beat of each half bar.

(An alternative would be to split the writes evenly between beats. But there are only ~6 in the worst case, which isn't very many. May as well batch them to take maximize NVMe utilization.)

sentientwaffle committed:
LSM/ManifestLog: Append to superblock manifest on block close

When there were multiple manifest logs, this needed to be done in `flush()` (synchronous with respect to the start of the bar) to avoid determinism issues as different trees raced.
But with a single manifest log, this ordering is guaranteed regardless – the append is more naturally part of `close_block()`.

sentientwaffle committed:
LSM/ManifestLog: Remove unnecessary flush() branch

sentientwaffle committed:
VSR/SuperBlockManifest: Fix broken test

This test is broken in main.
The break was detected thanks to the new `assert(address_removed)` in `SuperBlockManifest.remove()`.

eatonphil committed:
Merge pull request #878 from tigerbeetle/pe/cli2

Client built into the binary

eatonphil committed:
Use consistent relative path for TigerBeetle in README

eatonphil committed:
Merge pull request #1173 from tigerbeetle/pe/consistent-readme

Use consistent relative path for TigerBeetle in README

sentientwaffle committed:
LSM/Forest: Use inline-switch to dispatch tree ids with jump table

eatonphil committed:
No REPL printing in tests

cb22 committed:
Rename --enable-statsd to just --statsd, disable CI benchmark for now

sm2n committed:
lsm: in compaction, assert values_in is nonempty instead of branching

eatonphil committed:
Merge pull request #1175 from tigerbeetle/pe/repl-stdio

No REPL printing in tests

eatonphil committed:
Add --cluster=0 to tigerbeetle client example

sentientwaffle committed:
LSM: Code review

eatonphil committed:
Merge pull request #1176 from tigerbeetle/pe/more-readme-consistency

Add --cluster=0 to tigerbeetle client example

sentientwaffle committed:
VSR/Grid: Decouple block "fulfill" and "repair"

# Context

Big picture: This is all setup for proactive LSM-table sync.

Currently, the grid has a `Grid.read_global_queue` – a linked list of grid reads originating from the LSM (_not_ originating from a remote request).

When the grid receives a block over the network, it that block's address/checksum is queued in the `read_global_queue`, then:
- write that block to the grid on disk,
- and resolve the pending grid read(s).

# Table Sync

During table sync, we want to write the syncing blocks to disk, but there are (usually) no corresponding pending reads to resolve.

These two actions are decoupled:
1. The `Grid.read_global_queue` still exists, but only handles the "resolve the pending grid read(s)" case.
2. The `Grid.repair_queue` (a new data structure) tracks which blocks need to be repaired on disk.

When a `Grid.Read` is queued into `read_global_queue`, the address/checksum is also added to the `Grid.repair_queue` (assuming the latter has sufficient capacity).

# Optimizations

(TODO split this into separate commit?)

At checkpoint, the `Grid` does not need to wait for _all_ of its in-progress block writes to finish – only writes to blocks that are about to be released by the checkpoint.

cb22 committed:
Merge pull request #1166 from tigerbeetle/fix-ci-benchmark

Fix CI benchmark client args

sentientwaffle committed:
Storage/Test: Verify NextTick references are unique

sentientwaffle committed:
FIFO: Add contains()

sentientwaffle committed:
FIFO: Verify !contains on push()

sentientwaffle committed:
LSM/Forest: Allow tree_id_min > 1

sentientwaffle committed:
Merge pull request #1177 from tigerbeetle/dj-storage-next-tick-assertions

FIFO: Assert !contains(elem) before push(elem)

sentientwaffle committed:
LSM: Assertions and cleanup

matklad committed:
Merge pull request #1169 from sm2n/sm2n/table-cleanup

Some table cleanup

sentientwaffle committed:
LSM/Schema: ManifestLog: Var-arrays; tables then labels

Restructure the ManifestLog block schema:
- Instead of a fixed-length labels array, followed by the tables array, store a variable-length tables array followed by the labels array.
- Store the `entry_count` in the block header's `context` field.
- Order the tables array before the labels array since the alignment is higher.

We construct each ManifestLog block as if it is going to be full.
If it is not full (because it is the last block during a checkpoint) then copy the labels backwards.

sentientwaffle committed:
LSM/Schema: ManifestLog: Reorder labels/tables identifiers (refactor)

To match layout on disk.

sentientwaffle committed:
LSM/Schema: s/"manifest log block"/"manifest block"

We already had `BlockType.manifest`, this just makes the other terminology and identifiers consistent.

sentientwaffle committed:
LSM: Replace "pending" counter with bitset

sentientwaffle committed:
LSM/ManifestLog: Verify tree id is valid even for dropped table extents

sentientwaffle committed:
LSM/ManifestLog: Cleanup

- Rename `schema.Manifest.{labels,tables}_used()` to `{labels,tables}_const()`.
- Check `entry_count` against `size` when constructing `Manifest` block schema.
- Assert that grid blocks are non-empty.

sentientwaffle committed:
LSM/ForestTableIterator

sentientwaffle committed:
Fuzz/LSM: Prevent static checkpoint id

On checkpoint, we expect the checkpoint id to change.
We assert this in `VSRState.monotonic()`:

    assert(old.previous_checkpoint_id != new.previous_checkpoint_id);

The checkpoint id is `hash(hash(free_set), hash(manifest), hash(client_sessions))`.

Even if the LSM is "idle" for a whole log-wrap of messages (i.e. all lookups, no inserts/updates), then we still expect the client sessions to mutate – in order to checkpoint, we must have committed. And if we committed, then there must be new replies in our sessions.

The forest and tree fuzzers weren't mutating client sessions, though.

Tree fuzzer was failing on seed `6039089230838466239`.

sentientwaffle committed:
LSM/Forest: Clean up comptime loops

sentientwaffle committed:
Merge pull request #1178 from tigerbeetle/dj-lsm-fuzz-checkpoint-id

Fuzz/LSM: Prevent static checkpoint id

sentientwaffle committed:
VSR/Grid/RepairQueue: Add table-sync machinery

sentientwaffle committed:
VSR/Replica: Sync tables

# Context

State sync allows a replica to "skip" past a bunch of commits.

Immediately after [superblock sync](https://docs.tigerbeetle.com/internals/sync#algorithm) completes:

- The replica can participate in consensus (i.e. prepare/commit, view changes, etc).
- Manifest blocks are already synced (via normal [grid repair](https://docs.tigerbeetle.com/internals/vsr/#protocol-repair-grid)).
- The replica is missing the client replies that were skipped over. (These are repaired by the [sync client replies](https://docs.tigerbeetle.com/internals/vsr/#protocol-sync-client-replies) protocol.)
- The replica is missing the LSM table blocks (index/filter/data) that would have been created by the commits that were skipped over.

Previously TigerBeetle repaired the LSM table blocks lazily. That is, it goes about its normal business executing commits and compaction, and when it tries to read a block belonging to a skipped table, it interprets the missing block as corruption and repairs it.

This approach works (and is quite simple!), but has several downsides:

- This lazy approach is slow – the replica will often be blocked waiting for blocks from the network, and due to access patterns these delays will be sequential, not parallel.
- This lazy approach restores cluster durability incrementally in incompletely – there is no guarantee that the replica will eventually repair all missing blocks.

This changeset supplements (but does not replace!) this lazy approach with a proactive sync of skipped tables.

# Design

In pseudocode (see `Replica.sync_request_tables()`):

    for table in replica.state_machine.forest: # See ForestTableIterator.
        if table.snapshot_min ∈ (superblock.sync_op_min … superblock.sync_op_max):
            # Fetch the index block first.
            index_block := request table index block from cluster
            # Write it to disk:
            grid.repair_block(index_block)

            # Request and repair every block referenced by the table index block.
            for content in index_block:
                content_block := request content block from cluster
                grid.repair_block(content_block)

    # Then when all tables are fully repaired:
    superblock.sync_op_min = 0
    superblock.sync_op_max = 0

# Future

Right now this still leans on grid-repair protocol, requesting each block by address/checksum.

In the future we can require just a table index address/checksum along with a (compressed) bitset of which content blocks we require, and attach that to e.g. pings.

sentientwaffle committed:
VSR,LSM: Internal verification of grid blocks and tables

sentientwaffle committed:
VOPR: Don't require liveness when core is missing block

If the entire core of (non-standby) replicas is missing (and trying to repair) the same block, then the VOPR shouldn't signal it as a liveness bug.

sentientwaffle committed:
Test/VSR: Assert grids match after sync completes

sentientwaffle committed:
Test/StorageChecker: Check grid at checkpoints

sentientwaffle committed:
Merge pull request #1172 from tigerbeetle/dj-lsm-manifest-unified-2

LSM: Unified manifest

sentientwaffle committed:
VSR: Misc assertions and cleanup

sentientwaffle committed:
Docs: s/"sync grid protocol"/"sync tables protocol"

"Sync grid" is misleading since this protocol doesn't actually sync manifest blocks, only index/filter/data blocks.

matklad committed:
Apply suggestions from code review

Co-authored-by: djg <sentientwaffle@gmail.com>

matklad committed:
review comments

sentientwaffle committed:
CI: Escape wildcard "." when grepping version

sentientwaffle committed:
Merge pull request #1181 from tigerbeetle/dj-ci-wildcard

CI: Escape wildcard "." when grepping version

sentientwaffle committed:
VSR/GridRepairQueue: Code review cleanup/bugfixes

matklad committed:
style: state invariants positively

matklad committed:
Apply suggestions from code review

Co-authored-by: Joran Dirk Greef <joran@tigerbeetle.com>

matklad committed:
Hemming the way this is phrased

jorangreef committed:
Merge pull request #1188 from tigerbeetle/matklad/invariants

style: state invariants positively

sentientwaffle committed:
Test/GridChecker

Verify grid coherence across the cluster.

Specifically, on block write or (coherent) block read, verify that the address corresponds to at most one block version (i.e. checksum).

This extra check is disabled during replica divergence: The checkpoint id used for these checks is the "past" one, not the "future" (since the future is not known).

sentientwaffle committed:
VSR/GridRepairQueue: Poll to reclaim RepairTable's

Don't call `enqueue_table()` from within `checkpoint_commence()`.
(This worked and was safe, but is perhaps hard to follow.)

Instead, queue up the freed `RepairTable`s in the `GridRepairQueue`, and reclaim them by polling.

cb22 committed:
LSM: Small typo fix in manifest.zig

cb22 committed:
Merge pull request #1189 from tigerbeetle/cb22-typo-fix

LSM: Small typo fix in manifest.zig

matklad committed:
Apply suggestions from code review

Co-authored-by: Joran Dirk Greef <joran@tigerbeetle.com>

matklad committed:
docs: review fixes

matklad committed:
docs: explain how snapshot_max is initialized

matklad committed:
Merge pull request #1190 from tigerbeetle/matklad/snapshot-max

docs: explain how snapshot_max is initialized

sentientwaffle committed:
LSM/ForestTableIterator: Avoid O(trees) loop

Plus assertions and explanation.

sentientwaffle committed:
VSR/Sync: Remove unnecessary repair-table check

sentientwaffle committed:
VOPR: core_missing_blocks: Replica repairing a block it has

Allow a replica to get stuck repairing a block that is actually in its grid, if no other replica in the core has the block.

This is a combination of:
- We sync tables without checking that the blocks are present. (It would be a lot of mostly-unnecessary work otherwise.)
- If a replica state syncs from checkpoint X to checkpoint Y, it might have compacted some tables within checkpoint X. It syncs those anyway (for simplicity).

VOPR seed `901913452689623352`.

sentientwaffle committed:
Test/GridChecker: Move into Storage

To avoid runtime branch.

sentientwaffle committed:
VSR: Misc assertions

The removed assertions (in `Replica`) were redundant -- we already check `executing == 0`.

sentientwaffle committed:
VSR: s/GridRepairQueue/GridBlocksMissing

B/c table sync isn't logically repair (even though it uses the same protocol).

batiati committed:
Refactor Accounts and Transfers to u128 amounts

- Account balances fields are now u128.

- Transfers amounts are u128.

- Timeouts are u32 and expressed as seconds instead of nanoseconds.

- Three user_data fields, as u128, u64 and u32.

- AccountMutable and AccountImmutable are now a single groove Account.

- Reordered the Transfer and Account fields to keep @alignOf(T) == 16.

matklad committed:
Merge pull request #1152 from tigerbeetle/matklad/data-file-docs

docs: what we store on disk

sentientwaffle committed:
Forest: Order tree_infos by tree_id

sentientwaffle committed:
Test/StorageChecker: Add ChecksumStream to verify grid consistency

batiati committed:
Review

batiati committed:
Merge pull request #1157 from tigerbeetle/batiati-u128

Refactor Accounts and Transfers to u128 amounts

batiati committed:
Fix transfers.md

batiati committed:
Merge pull request #1192 from tigerbeetle/batiati-transfers-docs

Fix transfers.md

sentientwaffle committed:
Merge pull request #1179 from tigerbeetle/dj-sync-grid

VSR: State Sync (Part 4/?): Sync tables

NeilW committed:
Update fractional-amounts.md

Update 64 to 128 bit in the documentation

batiati committed:
Merge pull request #1194 from NeilW/patch-1

Update fractional-amounts.md

ReppCodes committed:
Add lookup return code enums

This commit addresses issue 1184.

Lookup REPL actions currently return hardcoded error strings. Create
enums for looking up accounts and transfers, and refactor the error
message to reference them.

The actual lookup functions still return Account and Transfer structs,
as a wrapper struct around them to also house the error code seemed like
overkill given there is currently only one error mode supported. This
could be added in the future if needed

sentientwaffle committed:
StateMachine: Prefetch from posted groove

Regression via https://github.com/tigerbeetle/tigerbeetle/pull/1157

sentientwaffle committed:
StateMachine: Add replica index to debug logs

sentientwaffle committed:
Merge pull request #1196 from tigerbeetle/dj-sm-prefetch-posted

StateMachine: Prefetch from posted groove

cb22 committed:
Node: fix flaky Amazon Linux test

cb22 committed:
Merge pull request #1197 from tigerbeetle/cb22/node-flake

Node: fix flaky Amazon Linux test

cb22 committed:
Docs: fix typo referencing create in lookup docs

cb22 committed:
Docs: an transfer -> a transfer

cb22 committed:
Merge pull request #1199 from tigerbeetle/cb22/docs-lookup

Docs: small typo and grammar fixes

cb22 committed:
Docs: Remove copyright notice from README

Chatted with @jorangreef, and we're removing these since:

1. They're not entirely accurate.
2. To make our README look more contributor friendly.

cb22 committed:
Merge pull request #1200 from tigerbeetle/cb22-update-readme

Docs: Remove copyright notice from README

cb22 committed:
Merge pull request #1195 from ReppCodes/lookup_enums

Add lookup return code enums

sentientwaffle committed:
Test/StorageChecker: Don't parameterize over Storage

(`StorageChecker` requires `TestStorage` anyway.)
(This is a big diff, but only because the `StorageChecker`'s methods are all un-indented.

sentientwaffle committed:
Grid: Verify grid block sector padding is zeroed

pragmaticanon committed:
Add eager parsing with clear indication of input

pragmaticanon committed:
merge main

pragmaticanon committed:
Update README with repl eager parsing

sentientwaffle committed:
VSR/Grid: Only write logical data, not padding sectors

sentientwaffle committed:
Merge pull request #1203 from tigerbeetle/dj-test-storage-checker-cleanup

Test/StorageChecker: Cleanup; verify grid block padding

cb22 commented:
Hey @pragmaticanon - thanks for the PR!

Do you think you could make it so `\` isn't a required delimiter, but it still eagerly parses on newline and aborts if there's any error? Maybe the newline could be special cased?

We're going for roughly the same behaviour as eg `psql` / `mysql`. (although, neither of those will give you an early indication something is wrong)

sentientwaffle committed:
VSR: Fix a couple storage bounds

The grid issue was hidden because our `block_size == sector_size` during tests.

sentientwaffle committed:
VSR/Replica: Fix outdated comment

sentientwaffle committed:
Merge pull request #1205 from tigerbeetle/dj-sector-fixes

VSR: Fix a couple of storage bounds

pragmaticanon committed:
Merge branch 'main' into repl_eager_parsing

pragmaticanon commented:
Sure yeah makes sense, will update
